### DISCUSSION TITLE
Proposal: deletedBytesCount for EC encoded volumes

### DISCUSSION THREAD

**Discussion by vrvico** [seaweedfs/seaweedfs#5761] [Ideas]
This discussion seeks to suggest a (hopefully simple) idea to improve management of ec-encoded data in a Seaweedfs cluster, with practical application as a public storage service (many customers writing and deleting many files). If an object is deleted from filer, it is not deleted from ec-encoded volume. But currently, the object is also not marked as deleted on the ec volume. The process during which free space could then be reclaimed is to first decode all volumes, then run fsck with -reallyDeleteFromVolume. Therefore as the cluster grows, it becomes exponentially slower to reclaim free space, to a point where it is no longer feasible to do it. This space then becomes lost. 

I propose for ec volumes to have the deletedBytesCount metric like normal volumes. This change would allow a cluster admin to define their own logic how to handle volumes with growing trash. It then becomes trivial for the admin to manage the free space - for example, a simple bash script cronjob that uses weed shell to iterate over ec volumes, pick those with garbage threshold reached, lock the cluster, decode volumes and run `volume.fsck -volumeId=$vol -reallyDeleteFromVolume` on them. Presumably after such operation the deletedBytesCount would be reset. The cluster could easily be managed with even hundreds of thousands of ec volumes.

**Comment by chrislusf** [comment]
deleted entry count would be much faster to get. Just need to sum up the `.ecj` file size across all servers.

**Reply by vrvico** [reply]
Deleted entry is indeed super efficient to get:
```
root@SFS-FSN1-DC1-0004 ~ # time bash ecj_size.sh
Device /media/sda1 -- Total .ecj files: 7799. Total .ecj files larger than 0 bytes: 0
Device /media/sdb1 -- Total .ecj files: 7220. Total .ecj files larger than 0 bytes: 0
Device /media/sdc1 -- Total .ecj files: 8011. Total .ecj files larger than 0 bytes: 0
Device /media/sdd1 -- Total .ecj files: 7813. Total .ecj files larger than 0 bytes: 0
Device /media/sde1 -- Total .ecj files: 6161. Total .ecj files larger than 0 bytes: 0
Device /media/sdf1 -- Total .ecj files: 7885. Total .ecj files larger than 0 bytes: 0
Device /media/sdg1 -- Total .ecj files: 6766. Total .ecj files larger than 0 bytes: 0
Device /media/sdh1 -- Total .ecj files: 7470. Total .ecj files larger than 0 bytes: 0
Device /media/sdi1 -- Total .ecj files: 7134. Total .ecj files larger than 0 bytes: 0
Device /media/sdj1 -- Total .ecj files: 7547. Total .ecj files larger than 0 bytes: 0

real	0m0.454s
user	0m0.294s
sys	0m0.159s
```
These are 15TB drives, half free. I am not sure why they return 0 bytes, probably this is something else to look at. But how would deleted entry count relate to the fact that one volume could have 1 deleted file that is 5GB, while another might have 200 small files deleted in total size 10MB? If set to trigger only based on number of deleted entries, the first volume would never trigger the process, I guess.
